### ISSUE TITLE
Creating a TTS extension

### ISSUE THREAD

**Issue by SwannSchilling** [oobabooga/text-generation-webui#72]
I wrote a simple extension which is sending the chat output my TTS endpoint and plays the wav file.
Is this the proper way of importing modules when writing extensions?

```python
import requests
import os
import winsound

params = {}

def play_tts():
    # Get the list of all wav files in the folder
    folder = 'path/to/folder'
    files = [f for f in os.listdir(folder) if f.endswith('.wav')]

    # Sort the files by their modification time
    files.sort(key=lambda x: os.path.getmtime(os.path.join(folder, x)))

    # Play the last modified wav file
    filename = os.path.join(folder, files[-1])
    winsound.PlaySound(filename, winsound.SND_FILENAME)

def remove_surrounded_chars(string):
    new_string = ""
    in_star = False
    for char in string:
        if char == '*':
            in_star = not in_star
        elif not in_star:
            new_string += char
    return new_string

def input_modifier(string):
    """
    This function is applied to your text inputs before
    they are fed into the model.
    """ 

    return string

def output_modifier(string):
    """
    This function is applied to the model outputs.
    """
    print(string)
    string = remove_surrounded_chars(string)
    # The API endpoint
    url = "http://192.168.2.114:8124/synthesize/"+string

    # A GET request to the API
    response = requests.get(url)
    play_tts()
    return string

def bot_prefix_modifier(string):
    """
    This function is only applied in chat mode. It modifies
    the prefix text for the Bot and can be used to bias its
    behavior.
    """

    return string

**Comment by oobabooga** [comment]
Yes it is. You can import whatever modules you like in your script.

If you don't mind, and out of curiosity, what tts model/engine are you using?

**Comment by SwannSchilling** [comment]
This is the one that I am using...
https://github.com/R2D2FISH/glados-tts

I have been thinking if it would be a good idea to have a Discord server for this project, so there would be a place to talk about tts, different LLMs and such?
Or are there already servers that you would recommend?

**Comment by oobabooga** [comment]
@SwannSchilling I am not active on discord. The good thing about using these github issue pages is that they are indexed, so other people can more easily find solutions to their problems.

I didn't know about glados-tts. It's a lot of fun and very easy to install. With some small modifications I turned your extension into this:

![nice](https://user-images.githubusercontent.com/112222186/218343623-f30a7a80-7fa7-4362-890e-4230c14fd3d0.png)

This is the source code:

```python
params = {}

def remove_surrounded_chars(string):
    new_string = ""
    in_star = False
    for char in string:
        if char == '*':
            in_star = not in_star
        elif not in_star:
            new_string += char
    return new_string

def input_modifier(string):
    """
    This function is applied to your text inputs before
    they are fed into the model.
    """

    return string

def output_modifier(string):
    """
    This function is applied to the model outputs.
    """

    string = remove_surrounded_chars(string)
    string = string.replace('"', '')
    string = string.replace('“', '')
    string = string.replace('\n', ' ')
    string = string.strip()

    ip = "192.168.1.1"
    string = f'<audio src="http://{ip}:8124/synthesize/%22{string}%22" controls></audio>'

    print(string)
    return string

def bot_prefix_modifier(string):
    """
    This function is only applied in chat mode. It modifies
    the prefix text for the Bot and can be used to bias its
    behavior.
    """

    return string

```

It works pretty well. Is it possible to use a voice other than glados?

**Comment by SwannSchilling** [comment]
That looks great!! And yes, I guess there are some ways to train other voices using Tacotron but I have not looked into this...
I will try to check for possible options and post them once I have found something!
There are a lot of models on https://uberduck.ai/ but there is unfortunately no option to download the pretrained ones!

**Comment by SwannSchilling** [comment]
Just found the guide to train your own voices here
https://uberduck-guide.gitbook.io/uberduck-written-tutorial/training-time/easy-or-hard

**Comment by oobabooga** [comment]
@SwannSchilling did you manage to use the Crust model? I seems like it is not compatible with the glados-tts code.

By the way, I forgot to mention that you should run the script above with the `--no-stream` option.

**Comment by Ph0rk0z** [comment]
Nice.. I was looking for a way to train some voices since elevenlabs wants $5 and ties it to your identity! We will have way better voices than CAI.

**Comment by SwannSchilling** [comment]
If I am putting the .pt to the models directory and am changing the model name in the tts engine script I am getting voice output...it does not improve the quality much, but to fully understand how this works I might have to take a deeper look into how tacotron2 and vocoders are working. I never trained my custom tts model before!
But the Uberduck Colab seems to be a good start!

Also Speechbrain on Huggingface might be a good place to look when trying to figure out how to implement tts...

**Comment by oobabooga** [comment]
Elevenlabs is spectacular, but it's a proprietary model hidden behind an API like gpt-3 and cai.

There is also tortoise-tts, but it uses a ton of VRAM and takes a minute to generate audio for a single phrase (on Colab at least), whereas glados-tts is almost instantaneous.

**Comment by oobabooga** [comment]
This seems like the ideal candidate: https://github.com/snakers4/silero-models

It has 117 voices in English and is easy to use:

```python
# V3
import torch

language = 'en'
model_id = 'v3_en'
sample_rate = 48000
speaker = 'en_0'
device = torch.device('cpu')

model, example_text = torch.hub.load(repo_or_dir='snakers4/silero-models',
                                     model='silero_tts',
                                     language=language,
                                     speaker=model_id)
model.to(device)  # gpu or cpu

audio = model.save_wav(text="I love language models, they are so useful.",
                        speaker=speaker,
                        sample_rate=sample_rate)

```

**Comment by SwannSchilling** [comment]
Awesome I will check this as soon as I am back home!!

**Comment by minipasila** [comment]
Would it be possible for it to also show the generated text with the voice?
Another thing I noticed is that if you use "Impersonate" it will just give the code for the media player.

**Comment by oobabooga** [comment]
@minipasila yes, with the --verbose option.

Impersonate is bugged at the moment when the last sent message was an image. I will create an issue for this.

**Comment by minipasila** [comment]
--verbose kind of works but it won't show the most recent generation, only the whole prompt that was entered.

**Comment by Ph0rk0z** [comment]
So what is the best system to use to locally clone voices from my own data with the least fuss.

**Comment by samiit** [comment]
> I have implemented a TTS extension using silero: https://github.com/oobabooga/text-generation-webui/tree/main/extensions/silero_tts
> 
> A total of 118 voices can be used. The default is `en_21`, but you can change it to anything between `en_0` and `en_117` and click on "Apply" to change the voice.
> 
> Here are samples for all the voices: https://oobabooga.github.io/silero-samples/

Hi,
Thank you for your great work. I wonder whether other languages that Silero supports - DE, RU, etc. can be also included, especially if you choose the chat to be in another language.
